### PR TITLE
WiP: NF: Metadata API proposal

### PR DESCRIPTION
**Pull request by hanke**
See #618 

I want to emphasize that this is not intended to be the one and only top-level command. It would still make sense to have a `describe` or analog functionality in `create` (or even `install`).

### CODE REVIEW

**Comment by bpoldrack** [comment]
Would you mind explaining, how you imagine the use of labels in #618 ?
I'm not sure, whether I correctly understand this idea. And I'd like to, before commenting on it ;)

**Comment by hanke** [comment]
Labels correspond to different metadata sets, and in the end different
files. If you look at the picture, we wanted to be able to import from
existing metadata and keep tracking changes. the easiest way to do that is
to keep the output separate from any other potential metadata (maybe
imported from somewhere else).

File: datalad/distribution/metadata.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            setting to manage multiple sets of metadata without impacting
+            the main (possible hand-curated) metadata."""),
+        aggregate=Parameter(
+            args=("--aggregate",),
+            action="store_true",
+            doc="""aggregate metadata from any installed subdataset. If a
+            recursion limit is set, installed subdataset exceeding this
+            limit will be ignored. Aggregation honors the selected metadata
+            category"""),
+        recursion_limit=recursion_limit,
```
Comment (yarikoptic) [inline datalad/distribution/metadata.py]: I wondered if we need an explicit `--recursive` to signal traversing the entire tree.  E.g. if we have ds1/ds2/ds3/ds4 hierarchy, and they all set (we need a persistent configuration setting for that) to collect meta only from direct child, but we want to "update" (i.e. aggregate) in all of them, while invoking this command from within ds1...

or may be such type of recursion could become a common option for more commands?  e.g. "save", and even "get" (after we introduce ability to request content e.g. by tags) could all have the same handling of such recursion -- just go from the leaves up (depth first)

File: datalad/distribution/metadata.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    1. Removal
+    2. Replacement
+    3. Addition
+    4. Aggregation
+    5. Query
+
+    Metadata items and terms can be addressed in various ways:
+
+    - by index (run `get *` to see items indexed): -> "#16"
```
Comment (yarikoptic) [inline datalad/distribution/metadata.py]: confused... what is `#16` here?

File: datalad/distribution/metadata.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    - by term index (run `get <someterm>` to see items indexed for this term):
+      -> "dc:contributor.author#4"
```
Comment (yarikoptic) [inline datalad/distribution/metadata.py]: Why `dc:` prefix?   if any custom them should be `datalad` or `dl`, I guess... where is the syntax with '#' separator  comes from?

File: datalad/distribution/metadata.py (hunk 2 above)
Comment (hanke) [inline datalad/distribution/metadata.py]: it refers to the 16th item in the list of meta data items, which you'd see from `get *`

File: datalad/distribution/metadata.py (hunk 3 above)
Comment (hanke) [inline datalad/distribution/metadata.py]: This is copied from Dryad. "dc:" stands for Dublin Cor, hence the opposite of "custom". "#" is the sugar that enable simple address for the n-th item in the list of matches.

File: datalad/distribution/metadata.py (hunk 1 above)
Comment (hanke) [inline datalad/distribution/metadata.py]: Sounds reasonable. Do you have a suggestions for a name?

Comment (hanke) [inline datalad/distribution/metadata.py]: What about `--apply-recursively`?

File: datalad/distribution/metadata.py (hunk 2 above)
Comment (yarikoptic) [inline datalad/distribution/metadata.py]: still confused -- is that a number of meta data items or index of the last one?

File: datalad/distribution/metadata.py (hunk 3 above)
Comment (yarikoptic) [inline datalad/distribution/metadata.py]: gy, forgot about the elephant ;) thanks

File: datalad/distribution/metadata.py (hunk 1 above)
Comment (yarikoptic) [inline datalad/distribution/metadata.py]: sounds reasonable

**Comment by hanke** [comment]
FTR: no longer relevant

**Pull request closed without merging**